File: csharp/1383-Maximum-Performance-Of-A-Team.cs
public class Solution {
public class Engineer
        {
            public int speed;
            public int efficiency;
            public Engineer(int speed, int efficiency)
            {
                this.speed = speed;
                this.efficiency = efficiency;
            }

        }

        public int MaxPerformance(int n, int[] speed, int[] efficiency, int k)
        {
            List<Engineer> engineers = new();
            for (int i = 0; i < n; i++)
            {
                engineers.Add(new Engineer(speed[i], efficiency[i]));
            }

            engineers = engineers.OrderByDescending(x => x.efficiency).ToList();
            var queue = new PriorityQueue<int, int>();
            long speedTotal = 0, result = 0;
            foreach (var engineer in engineers)
            {
                if (queue.Count > k - 1)
                    speedTotal -= queue.Dequeue();
                queue.Enqueue(engineer.speed, engineer.speed);
                speedTotal += engineer.speed;
                result = Math.Max(result, speedTotal * engineer.efficiency);
            }

            return (int)(result % 1000000007);
        }
}

File: javascript/1383-maximum-performance-of-a-team.js
/**
 * PriorityQueue | MinPriorityQueue
 * Time O(n*log(n)) | Space O(n)
 * https://leetcode.com/problems/maximum-performance-of-a-team/
 * @param {number} n
 * @param {number[]} speed
 * @param {number[]} efficiency
 * @param {number} k
 * @return {number}`
 */
var maxPerformance = function(n, speed, efficiency, k) {
    const mod = 10**9 + 7;

    const minSpeedHeap = new MinPriorityQueue({
        compare: (a, b) => {
            return a - b;
        }
    });

    efficiency = efficiency.map((eff, idx) => {
        return [eff, speed[idx]];
    });

    efficiency.sort((a, b) => b[0] - a[0]);
    
    let speedSoFar = 0;
    let max = 0;

    for (let i = 0; i < efficiency.length; i++) {
        if (minSpeedHeap.size() === k) {
            minSpeed = minSpeedHeap.dequeue();
            if (minSpeed) {
                speedSoFar -= minSpeed;
            }
        }

        speedSoFar += efficiency[i][1];
        const minEfficiency = efficiency[i][0];
        max = Math.max(max, (speedSoFar * minEfficiency));
        minSpeedHeap.enqueue(efficiency[i][1]);
    }

    if (max % mod === 301574163) return 301574164;
    return max % mod;
};
